SourceCode/GPS/Classes/CISOBUS.cs

```diff
@@ -0,0 +1,157 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace AgOpenGPS
+{
+    public class CISOBUS
+    {
+        private readonly FormGPS mf;
+
+        private bool wasAlive;
+        private long timestamp;
+
+        private bool sectionControlEnabled;
+        private List<bool> actualSectionStates = new List<bool>();
+
+        public CISOBUS(FormGPS _f)
+        {
+            //constructor
+            mf = _f;
+        }
+
+        public bool IsSectionOn(int section)
+        {
+            if (section < actualSectionStates.Count)
+            {
+                return actualSectionStates[section];
+            }
+            return false;
+        }
+
+        public void RequestSectionControlEnabled(bool enabled)
+        {
+            // Send the request
+            byte[] data = new byte[7];
+            data[0] = 0x80; // standard AIO header
+            data[1] = 0x81; // PGN header
+            data[2] = 0x7F; // SRC address
+            data[3] = 0xF1; // PGN
+            data[4] = 0x02; // Length
+            data[5] = (byte)(enabled ? 0x01 : 0x00); // Section control enabled request
+            mf.SendPgnToLoop(data);
+        }
+
+        public bool IsSectionControlEnabled()
+        {
+            return sectionControlEnabled;
+        }
+
+        private void SetSectionControlEnabled(bool enabled)
+        {
+            if (sectionControlEnabled != enabled)
+            {
+                // Changed, act accordingly
+                sectionControlEnabled = enabled;
+
+                if (sectionControlEnabled)
+                {
+                    mf.btnIsobusSC.Image = Properties.Resources.IsobusSectionControlOn;
+                }
+                else // Section control disabled
+                {
+                    mf.btnIsobusSC.Image = Properties.Resources.IsobusSectionControlOff;
+                }
+            }
+        }
+
+        public bool IsAlive()
+        {
+            // Check if the timestamp is not older than 1 second
+            bool isAlive = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() - timestamp < 1000;
+
+     
+            if (wasAlive && !isAlive)
+            {
+                // Lost connection
+                mf.btnIsobusSC.Visible = false;
+            }
+            else if (!wasAlive && isAlive)
+            {
+                // Reconnected
+                mf.btnIsobusSC.Visible = true;
+            }
+
+            wasAlive = isAlive;
+            return isAlive;
+        }
```

Comment:
We determine if an ISOBUS implement is connected by checking if we received an heartbeat within the last second. This heartbeat contains all the information AOG needs to correctly display the current state of the TC.

Reply:
```suggestion
        public bool IsAlive()
        {
            // Check if the timestamp is not older than 1 second
            bool isAlive = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() - timestamp < 1000;

            mf.btnIsobusSC.Visible = isAlive;

            wasAlive = isAlive;
            return isAlive;
        }
```

Reply:
public bool IsAlive()
        {
            // Check if the timestamp is not older than 1 second
            bool isAlive = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() - timestamp < 1000;
```suggestion
           mf.btnIsobusSC.Visible = isAlive;
```
            wasAlive = isAlive;
            return isAlive;
        }

Reply:
Seems that visibility can be set like this
mf.btnIsobusSC.Visible = isAlive

without doing any tests